Magic extended attributes rework + add 'chunk_list' magic xattr

janpriessnitz opened the pull request.

janpriessnitz commented:
@cernvm-bot cloudtest

janpriessnitz commented:
@jblomer this one is ready for your review.

jblomer commented:
@cernvm-bot cloudtest

jblomer commented on cvmfs/magic_xattr.h:
```diff
@@ -0,0 +1,251 @@
+};
+
+class AuthZMagicXattr : public BaseMagicXattr {
+  virtual bool PrepareValueFenced();
```
Doesn't this render the two virtual methods private, i.e. don't we need the `public` keyword?

janpriessnitz replied:
Good question! I haven't thought of that. I compared the magic xattr outputs between the old and the new version and it matched. So I guess it works as expected and override functions inherit the visibility as well.

jblomer commented on cvmfs/mountpoint.cc:
```diff
@@ -1167,6 +1167,8 @@ MountPoint *MountPoint::Create(
   mountpoint->CreateTables();
   mountpoint->SetupBehavior();
 
+  mountpoint->magic_xattr_mgr_ = new MagicXattrManager(mountpoint.weak_ref());
```
Let's put that in `SetupBehavior`.  The `hide_magic_xattrs_` flag can be incorporated in the `MagicXattrManager`

jblomer commented on cvmfs/magic_xattr.cc:
```diff
@@ -0,0 +1,569 @@
+#include "quota.h"
+#include "signature.h"
+
+#include "magic_xattr.h"
```
Move up as the first include

jblomer commented on cvmfs/magic_xattr.cc:
```diff
@@ -0,0 +1,569 @@
+  std::string result;
+  std::map<std::string, BaseMagicXattr *>::iterator it = base_xattrs_.begin();
+  for (; it != base_xattrs_.end(); ++it) {
+    result += (*it).first + std::string("\0", 1);
```
Perhaps better `result.push_back('\0')` to avoid one more string construction.

jblomer commented on cvmfs/magic_xattr.cc:
```diff
@@ -0,0 +1,569 @@
+    return NULL;
+  }
+
+  result->path_ = path;
```
Hm, here we have a thread-safety issue.  For the mountpoint it's easy, it is fixed for the lifetime of the manager, so in fact the magic xattr objects can get the mountoint on construction.

Path and dirent need to be locked somehow.  I think the magic xattr base class can have `Lock(path, dirent)` and `Release()` methods that make sure that path and dirent stay stable for `PrepareValueFenced()` followed by `Get()`.

The `MagicXattrManager::Get()` can then return an RAII wrapper around the xattr object that calls `Release()` on destruction.

jblomer commented on cvmfs/magic_xattr.cc:
```diff
@@ -0,0 +1,569 @@
+void MagicXattrManager::Register(const std::string &name,
+                                 BaseMagicXattr *magic_xattr)
+{
+  base_xattrs_[name] = magic_xattr;
```
We should make sure the name does not exist yet.

janpriessnitz commented:
@jblomer I've addressed your comments. I have cleaned up the commit sequence, but if you want to see what changed, compare this branch against janpriessnitz/magic-xattr-backup branch.

jblomer commented:
@cernvm-bot cloudtest

jblomer commented on cvmfs/magic_xattr.h:
```diff
@@ -0,0 +1,324 @@
+  virtual std::string GetValue();
+};
+
+class RawlinkMagicXattr : public BaseMagicXattr {
```
For consistency, this should inherit from a `SymlinkMagicXattr` class, i.e. the class hierarchy should mirror the flavor enum.

jblomer commented on cvmfs/magic_xattr.h:
```diff
@@ -0,0 +1,324 @@
+ */
+class MagicXattrManager {
+ public:
+  explicit MagicXattrManager(MountPoint *mountpoint, bool hide_magic_xattrs);
```
With two arguments, we can drop the `explicit`

jblomer commented on cvmfs/magic_xattr.h:
```diff
@@ -0,0 +1,324 @@
+
+class MountPoint;
+
+enum MagicXattrFlavor {
```
Let's add a comment explaining the relationship of this enum with the MagicXattr class hierarchy

jblomer commented on cvmfs/magic_xattr.h:
```diff
@@ -0,0 +1,324 @@
+  inline bool IsNull() const { return ref_ == NULL; }
+
+ protected:
+  BaseMagicXattr *ref_;
```
Perhaps better named `ptr_`?

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1880,6 +1609,70 @@ string *g_boot_error = NULL;
+};
+
+class InodeMaxMagicXattr : public BaseMagicXattr {
+  virtual bool PrepareValueFenced() { return true; }
```
Can be dropped.

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1880,6 +1609,70 @@ string *g_boot_error = NULL;
+};
+
+class MaxFdMagicXattr : public BaseMagicXattr {
+  virtual bool PrepareValueFenced() { return true; }
```
Can be dropped

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1880,6 +1609,70 @@ string *g_boot_error = NULL;
+};
+
+class PidMagicXattr : public BaseMagicXattr {
+  virtual bool PrepareValueFenced() { return true; }
```
Can be dropped

jblomer commented on cvmfs/cvmfs.cc:
```diff
@@ -1880,6 +1609,70 @@ string *g_boot_error = NULL;
+};
+
+class UptimeMagicXattr : public BaseMagicXattr {
+  virtual bool PrepareValueFenced() { return true; }
```
Can be dropped

jblomer commented on cvmfs/magic_xattr.cc:
```diff
@@ -0,0 +1,515 @@
+  for (; it != xattr_list_.end(); ++it) {
+    MagicXattrFlavor flavor = (*it).second->GetXattrFlavor();
+    // Skip those which should not be displayed
+    switch (flavor) {
```
Let's add a default which aborts to make sure we don't forget any flavor.

jblomer commented on cvmfs/magic_xattr.cc:
```diff
@@ -0,0 +1,515 @@
+    return MagicXattrRAIIWrapper(NULL);
+  }
+
+  result->mount_point_ = mount_point_;
```
We can set this on construction / in `Register()`

jblomer commented on cvmfs/magic_xattr.cc:
```diff
@@ -0,0 +1,515 @@
+  : mount_point_(mountpoint),
+    hide_magic_xattrs_(hide_magic_xattrs)
+{
+  xattr_list_["user.catalog_counters"] = new CatalogCountersMagicXattr();
```
Let's use our own `Register()` (e.g. to set the mountpoint)

jblomer commented on cvmfs/magic_xattr.h:
```diff
@@ -0,0 +1,324 @@
+ *    Instead, member variables are set and the original instance is returned.
+ *    A mutex prevents from race conditions in case of concurrent access.
+ */
+class MagicXattrManager {
```
That should inherit from `SingleCopy` because it managed the magic xattr objects.

jblomer commented on cvmfs/magic_xattr.h:
```diff
@@ -0,0 +1,324 @@
+class MagicXattrRAIIWrapper {
+ public:
+  inline MagicXattrRAIIWrapper() : ref_(NULL) { }
+  inline explicit MagicXattrRAIIWrapper(BaseMagicXattr *ref) : ref_(ref) { }
```
I think for symmetry reasons, we should call `Lock()` in the RAII constructor

jblomer commented on cvmfs/magic_xattr.cc:
```diff
@@ -0,0 +1,515 @@
+
+std::string ExternalTimeoutMagicXattr::GetValue() {
+  unsigned seconds, seconds_direct;
+  mount_point_->download_mgr()->GetTimeout(&seconds, &seconds_direct);
```
This needs to be the external download manager

jblomer commented on cvmfs/magic_xattr.cc:
```diff
@@ -0,0 +1,515 @@
+}
+
+bool NIOErrMagicXattr::PrepareValueFenced() {
+  n_io_err_ = mount_point_->file_system()->n_io_error()->ToString();
```
We can move this to `GetValue()`, can't we?

jblomer commented on cvmfs/magic_xattr.cc:
```diff
@@ -0,0 +1,515 @@
+}
+
+bool UsedDirPMagicXattr::PrepareValueFenced() {
+  n_used_dirp_ = mount_point_->file_system()->no_open_dirs()->ToString();
```
Can be moved to `GetValue`, I think

jblomer commented on cvmfs/magic_xattr.cc:
```diff
@@ -0,0 +1,515 @@
+}
+
+bool UsedFdMagicXattr::PrepareValueFenced() {
+  n_used_fd_ = mount_point_->file_system()->no_open_files()->ToString();
```
Can be moved to `GetValue`, I think.

jblomer commented:
@cernvm-bot cloudtest cc7
